pkg/web/apps/apps_service.go

```diff
@@ -79,3 +81,103 @@ func (a *appsService) DisableAllAppVersionsByAppID(id string, message string) er
 
 	return a.repository.DisableAllAppVersionsByAppID(app.AppID, message)
 }
+
+// // InitClientApp updates the number of
+func (a *appsService) InitClientApp(deviceInfo *models.Device) (*models.Version, error) {
+	if _, err := a.repository.GetAppByAppID(deviceInfo.AppID); err != nil {
+		return nil, err
+	}
+
+	version, err := a.repository.GetVersionByAppIDAndVersion(deviceInfo.AppID, deviceInfo.Version)
+
+	// If any error other Not Found error occurred, return
+	if err != nil && err != models.ErrNotFound {
+		return nil, err
+	}
+
+	// If the version does not exist, create it
+	if err == models.ErrNotFound {
+		version = &models.Version{
+			ID:      uuid.New().String(),
+			Version: deviceInfo.Version,
+			AppID:   deviceInfo.AppID,
+		}
+	}
+
+	// Increment the number of app launches
+	version.NumOfAppLaunches++
+
+	// Update the existing version or create a new one
+	if err := a.repository.UpsertVersionWithAppLaunchesAndLastLaunched(version); err != nil {
+		return nil, err
+	}
+
+	device, err := a.repository.GetDeviceByDeviceIDAndAppID(deviceInfo.DeviceID, deviceInfo.AppID)
+
+	if err != nil {
+		// If we can't find the device by device ID and app ID
+		if err != models.ErrNotFound {
+			return nil, err
+		}
+		// Build a new device to save to the database
+		device = newDeviceFromDeviceAndVersion(*deviceInfo, *version)
+	}
+
+	if updateDeviceVersionID(device, version.ID) || updateDeviceDeviceVersion(device, deviceInfo.DeviceVersion) {
+
+		err := a.repository.InsertDeviceOrUpdateVersionID(*device)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// clear these values before returning the data
+	version.LastLaunchedAt = ""
+	version.NumOfAppLaunches = 0
+
+	return version, nil
+}
+
+// If the Device.VersionID property is different to @var versionID,
+// update it to match, returning a bool to indicate if it was updated
+func updateDeviceVersionID(device *models.Device, versionID string) bool {
+	var isUpdated bool
+
+	if device.VersionID != versionID {
+		device.VersionID = versionID
+
+		isUpdated = true
+	}
+
+	return isUpdated
+}
+
+// If the Device.DeviceVersion property is different to @var deviceVersion,
+// update it to match, returning a bool to indicate if it was updated
+func updateDeviceDeviceVersion(device *models.Device, deviceVersion string) bool {
+	var isUpdated bool
+
+	if device.DeviceVersion != deviceVersion {
+		device.DeviceVersion = deviceVersion
+
+		isUpdated = true
+	}
+
+	return isUpdated
+}
+
+// Build a new Device model from a combination of Device and Version model properties.
+func newDeviceFromDeviceAndVersion(d models.Device, v models.Version) *models.Device {
```

Comment:
Who knows more one entity is the proper entity. I'd like to recommend you add in the models its constructors.In this way, in the model.device you will have a constructor as in the version you will have another. Also, adding the constructors in the models allow it to be used in other places.

Reply:
Look for how to create constructors. see [here](https://golangbot.com/structs-instead-of-classes/) for example.

Reply:
Yes I moved this into a ctor. Thanks for the suggestion.

Reply:
I am closing this one as resolved so.